### DISCUSSION TITLE
표준 예외와 커스텀 예외의 장단점을 말해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#91] [Java]
.

**Comment by jimi567** [comment]
표준 예외의 장점으로는 웬만한 java 개발자라면 익숙하기 때문에 다른 개발자들이 어떤 예외 상황인지 파악하기가 쉽다.
하지만, 상황에 따라 적합한 예외가 존재하지 않을 수 있다는 단점이 있고, 상위에서 한 번 catch할 경우, 수많은 예외 상황에서 특정 예외 상황을 걸러내기가 어렵다.

커스텀 예외를 활용하게 된다면, 다음 장점이 있다.
클래스 이름만으로 의미전달이 가능하다.
클래스 내부에서 특별한 후처리가 가능하다.
표준예외가 문자열만을 저장하는 것과는 달리, 커스텀 예외는 필요한 객체들을 저장하여 사용할 수 있다.
커스텀 예외는 Throwable의 메서드인 fillInStackTrace를 오버라이드해 stack trace를 제어할 수 있다.(간략화 또는 출력 자체를 x)

단 커스텀 예외는 관리 클래스가 많아지고, 이름만으로 이해가 되지 않을 경우 동료 개발자에게 파악을 위한 리소스를 야기할 수 있다.

**Comment by nayonsoso** [comment]
표준 예외는 동료 개발자들에게도 익숙한 예외이기 때문에 사용하기 쉽고, 의미 전달이 잘 됩니다. 
대표적인 표준 예외들로는 
- 부적절한 값이 들어올 때 던지는 `IllegalArgumentException`
- 일을 수행하기에 적합하지 않은 상태일 때 던지는 `IllegalStateException`
- 요청받은 작업을 지원하지 않는 경우에 던지는`UnsupportedOperationException`

들이 있는데요, 이런 예외는 모두에게 익숙하기 때문에 쓰임을 바로 파악할 수 있습니다.

반면 커스텀 예외의 경우, 해당 커스텀 예외가 언제 발생하는지 이해해야 합니다.
또한 커스텀 예외가 많아질 경우 클래스의 갯수가 많아질 수 있다는 단점이 있습니다.
하지만 표준 예외로 전달하기 어려운 정보를 네이밍으로 전달할 수 있다는 장점 또한 있습니다.

cf.
개인적으로는 표준 예외를 선호합니다.
추가 정보는 예외 메세지로 전달할 수 있다고 생각하기 때문입니다 ㅎㅎ
